Feature/wt 9718 tag property crud support open api spec

chanddan opened the pull request.

Commit:
[WT-9718] First cut of openapi tag property.

Commit:
[WT-9718] Tweaked openapi tag property.

Commit:
[WT-9718] Added tag managed by field.

Commit:
[WT-9718] Updated tag managed description.

tom-haines commented on spec/openapi.yaml:
```diff
@@ -458,6 +514,40 @@ components:
+            type: string
+      required:
+        - tagName
+        - tagProperties
```
Should be `propertyNames` ?

tom-haines commented on spec/openapi.yaml:
```diff
@@ -482,6 +572,18 @@ components:
+          type: string
+        tagProperties:
+          description: >
+            Add one or more tag properties to the tag. The default behaviour ensures that there will be **no** duplicate or deletion of any existing tag specific properties within WiseTime.
```
```suggestion
            You can assign metadata to a tag.  You provide this optional information as a name-value (key-value) pair.  In the case of providing any name-value where the name is already assigned as metadata to the tag, the value will be updated with value supplied.
```

tom-haines commented on spec/openapi.yaml:
```diff
@@ -482,6 +572,18 @@ components:
+        tagProperties:
+          description: >
+            Add one or more tag properties to the tag. The default behaviour ensures that there will be **no** duplicate or deletion of any existing tag specific properties within WiseTime.
+            Use the endpoint `/tag/properties/delete` for the deletion of tag properties.
```
```suggestion
            To delete/remove a name-value pair stored as metadata on a tag, you must use the `delete-tag-properties` operation.
```

tom-haines commented on spec/openapi.yaml:
```diff
@@ -482,6 +572,18 @@ components:
+            Identifier of the resource that manages this tag.
+            If blank or missing, the value will be defaulted to the connector's name.
+          type: string
+        tagProperties:
```
Suggest we follow naming convention of metadata, which looks more common name for this type of key-value map (others call it a 'tag' like AWS S3 API, which would be quite confusing in our context).

a. As we reuse same object in `AddSetTagPropertiesRequest` also, could we schema ref to a `TagMetadataSet` ?  We could then use schema ref also in schema rep of Tag itself as needed.

b. Assuming type is TagMetadataSet, could we then `tagProperties` to simply `metadata` ? 

c. Think might be clearer if we use `/tag/metadata` in place of `/tag/properties` as a tag has other properties, such as last modified timestamp and createdBy etc.

Property

tom-haines commented on spec/openapi.yaml:
```diff
@@ -458,6 +514,40 @@ components:
+          type: string
+          description: The tag from which to delete the tag property set.
+        propertyNames:
+          description: The tag property to delete.
```
```suggestion
          description: The list of tag property names to delete.
```

tom-haines commented on spec/openapi.yaml:
```diff
@@ -458,6 +514,40 @@ components:
+          additionalProperties:
+            type: string
+          example:
+            PRIMARY_GROUP: billable/non-billable
```
we should use lowercase in example, and mention in MetadataSet (see other comment) that all tag property names are converted/stored in lowercase.

tom-haines reviewed (changes requested):
Have made some initial suggestions re naming and schema ref reuse (not sure if we can or not)

chanddan committed:
Update spec/openapi.yaml

Co-Authored-By: Tom Haines <thomas.haines@gmail.com>

chanddan committed:
Update spec/openapi.yaml

Co-Authored-By: Tom Haines <thomas.haines@gmail.com>

chanddan committed:
Update spec/openapi.yaml

Co-Authored-By: Tom Haines <thomas.haines@gmail.com>

Commit:
[WT-9718] Changes based on PR suggestions.

chanddan commented:
@tom-haines I have updated the spec, based on your naming suggestion I have renamed "tag properties" to "metadata". Agreed, it's a more descriptive name for its intent.

Also added TagMetadataSet.

Commit:
[WT-9718] Changes based on PR suggestions.

tom-haines commented on spec/openapi.yaml:
```diff
@@ -120,6 +120,65 @@ paths:
 
+  /tag/metadata:
+    post:
+      operationId: add-set-tag-metadata
```
for clarity, suggest `set-tag-metadata` for this operation id

Commit:
[WT-9718] Renamed "addSet" to "update".

tom-haines commented on spec/openapi.yaml:
```diff
@@ -120,6 +120,65 @@ paths:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/UpdateTagMetadataRequest'
```
small nit - if operation called `Assign` then schema object should use same verb (replacing of `Update`), or vice-versa

tom-haines commented on spec/openapi.yaml:
```diff
@@ -120,6 +120,65 @@ paths:
 
+  /tag/metadata:
+    post:
+      operationId: assign-tag-metadata
```
nit: other parts of api use highest concept to lowest concept semantics, to this should be

`tag-metadata-udpate`, `TagMetadataUpdateRequest` etc

tom-haines reviewed (changes requested):
as discussed, let's follow concept ordering semantics for consistency

Commit:
[WT-9718] Change based on suggestion.

Commit:
[WT-9718] Changed to use "highest concept".

chanddan commented:
@tom-haines  I have resolved your nits.

Commit:
[WT-9718] Removed the need for tag-metadata-update from openapi spec.

Commit:
[WT-9718] Added 204 response to tag/metadata/delete.

tom-haines committed:
WT-9718 simplify language

tom-haines commented on spec/openapi.yaml:
```diff
@@ -120,6 +120,39 @@ paths:
+                $ref: '#/components/schemas/TagMetadataDeleteResponse'
+        '204':
+          description: >
+            (No Content) when there is no need to return any content, where metadata was not found or if it was already deleted.
```
Sorry to say but this design is still problematic, as we allow multiple `metadataNames` to be provided in the delete request, but only allow for a unary response.  I suggest we discuss to find clean solution.

vyshane commented on spec/openapi.yaml:
```diff
@@ -120,6 +120,39 @@ paths:
+                $ref: '#/components/schemas/TagMetadataDeleteResponse'
+        '204':
+          description: >
+            (No Content) metadata was not found or already deleted.
```
```suggestion
```

@chanddan I didn't realise that we were deleting multiple tag metadata entries in one request. If that is the case then it seems simpler and less confusing to always return 200 OK when tag was found and there was no error processing the request, whether metadata was deleted or not. My 204 suggestion doesn't seem all that great now. What do you think?

chanddan replied:
I have added a reply in `connect-api-server/pull-requests/159/overview`

vyshane commented on spec/openapi.yaml:
```diff
@@ -120,6 +120,39 @@ paths:
+              $ref: '#/components/schemas/TagMetadataDeleteRequest'
+      responses:
+        '200':
+          description: Tag metadata successfully deleted.
```
```suggestion
          description: Tag metadata successfully deleted. This operation is idempotent, and will also return HTTP code 200 if some or all of the tag metadata names were not found.
```

If you agree with [this suggestion](https://github.com/wisetime-io/connect-api-spec/pull/29/files#r396850452), then perhaps something like the above makes sense.

chanddan replied:
Removed 204 from spec.

Commit:
Merge branch 'master' of https://github.com/wisetime-io/connect-api-spec into feature/WT-9718_tag_property_CRUD_support_OpenAPI_spec

Commit:
Merge branch 'feature/WT-9718_tag_property_CRUD_support_OpenAPI_spec' of https://github.com/wisetime-io/connect-api-spec into feature/WT-9718_tag_property_CRUD_support_OpenAPI_spec

Commit:
[WT-9718] Removed 204 response for tagMetadataDelete.

Commit:
[WT-9718] Fixed typo.

tom-haines reviewed (approved):
On deployment of additional endpoint, this can be merged to master.
